full-moon/src/ast/parsers.rs

```diff
@@ -1663,129 +1995,1245 @@ fn parse_function_body(state: &mut ParserState) -> ParserResult<FunctionBody> {
+    state: &mut ParserState,
+    if_token: TokenReference,
+) -> Result<ast::IfExpression, ()> {
+    let ParserResult::Value(condition) = parse_expression(state) else {
+        return Err(());
+    };
 
-fn parse_name(state: &mut ParserState) -> ParserResult<Name> {
-    let Ok(current_token) =  state.current() else {
-        return ParserResult::NotFound;
+    let Some(then_token) = state.require(Symbol::Then, "expected `then` after condition") else {
+        return Err(());
     };
 
-    match current_token.token_type() {
-        TokenType::Identifier { .. } => {
-            let name_token = state.consume().unwrap();
-            ParserResult::Value(force_name(state, name_token))
-        }
+    let ParserResult::Value(if_expression) = parse_expression(state) else {
+        return Err(());
+    };
 
-        _ => ParserResult::NotFound,
+    let mut else_if_expressions = Vec::new();
+    while let Some(else_if_token) = state.consume_if(Symbol::ElseIf) {
+        let ParserResult::Value(condition) = parse_expression(state) else {
+            return Err(());
+        };
+
+        let Some(then_token) = state.require(Symbol::Then, "expected `then` after condition")
+        else {
+            return Err(());
+        };
+        let ParserResult::Value(expression) = parse_expression(state) else {
+            return Err(());
+        };
+        else_if_expressions.push(ast::ElseIfExpression {
+            else_if_token,
+            condition,
+            then_token,
+            expression,
+        })
     }
-}
 
-fn parse_name_with_attributes(state: &mut ParserState) -> ParserResult<Name> {
-    let Ok(current_token) =  state.current() else {
-        return ParserResult::NotFound;
+    let Some(else_token) = state.require(Symbol::Else, "expected `else` to end if-else expression")
+    else {
+        return Err(());
     };
 
-    match current_token.token_type() {
-        TokenType::Identifier { .. } => {
-            let name_token = state.consume().unwrap();
-            ParserResult::Value(force_name_with_attributes(state, name_token))
-        }
+    let ParserResult::Value(else_expression) = parse_expression(state) else {
+        return Err(());
+    };
 
-        _ => ParserResult::NotFound,
-    }
+    Ok(ast::IfExpression {
+        if_token,
+        condition: Box::new(condition),
+        then_token,
+        if_expression: Box::new(if_expression),
+        else_if_expressions: if else_if_expressions.is_empty() {
+            None
+        } else {
+            Some(else_if_expressions)
+        },
+        else_token,
+        else_expression: Box::new(else_expression),
+    })
 }
 
-fn force_name(_state: &mut ParserState, name: TokenReference) -> Name {
-    Name {
-        name,
-        #[cfg(feature = "lua54")]
-        attribute: None,
-    }
-}
+#[cfg(feature = "luau")]
+fn expect_interpolated_string(
+    state: &mut ParserState,
+    mut current: TokenReference,
+) -> Result<ast::InterpolatedString, ()> {
```

Comment:
Does this need to return Result? It looks like all codepaths return `Ok`.

Reply:
Good spot, looks unnecessary